File: android/app/src/main/java/com/pms_parkin_mobile/service/UserDataSingleton.java
package com.pms_parkin_mobile.service;

import android.content.Context;
import android.content.SharedPreferences;

import com.pms_parkin_mobile.dto.LobbyOpenData;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

public class UserDataSingleton {

    private String Dong;
    private String Ho;
    private String UserName;
    private String Cel;
    private String ID;
    private boolean IsDriver;
    private String userId;
    private ArrayList<LobbyOpenData> OpenMINOR;

    private SharedPreferences sharedPreferences;
    private static final String PREF_NAME = "user_data_pref";

    private static final UserDataSingleton ourInstance = new UserDataSingleton();

    public static UserDataSingleton getInstance() {
        return ourInstance;
    }

    private UserDataSingleton() {
        OpenMINOR = new ArrayList<>();
    }

    /** App.init() 에서 호출 — 프로세스 재시작 후에도 데이터 복원 */
    public void init(Context context) {
        sharedPreferences = context.getApplicationContext()
                .getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        loadFromPrefs();
    }

    private void loadFromPrefs() {
        if (sharedPreferences == null) return;
        Dong     = sharedPreferences.getString("dong",      null);
        Ho       = sharedPreferences.getString("ho",        null);
        UserName = sharedPreferences.getString("user_name", null);
        Cel      = sharedPreferences.getString("cel",       null);
        ID       = sharedPreferences.getString("id",        null);
        IsDriver = sharedPreferences.getBoolean("is_driver", false);
        userId   = sharedPreferences.getString("user_id",   null);

        String jsonStr = sharedPreferences.getString("open_minor", null);
        if (jsonStr != null) {
            try {
                OpenMINOR = new ArrayList<>();
                JSONArray arr = new JSONArray(jsonStr);
                for (int i = 0; i < arr.length(); i++) {
                    JSONObject obj = arr.getJSONObject(i);
                    LobbyOpenData data = new LobbyOpenData();
                    data.setMinor(obj.optString("minor"));
                    data.setRssi(obj.optString("rssi"));
                    data.setId(obj.optString("id"));
                    data.setDong(obj.optString("dong"));
                    data.setHo(obj.optString("ho"));
                    OpenMINOR.add(data);
                }
            } catch (JSONException e) {
                OpenMINOR = new ArrayList<>();
            }
        }
    }

    // ======== Getters ========
    public String getDong()                        { return Dong; }
    public String getHo()                          { return Ho; }
    public String getUserName()                    { return UserName; }
    public String getCel()                         { return Cel; }
    public String getID()                          { return ID; }
    public boolean isIsDriver()                    { return IsDriver; }
    public String getUserId()                      { return userId; }
    public ArrayList<LobbyOpenData> getOpenMINOR() { return OpenMINOR; }

    // ======== Setters (SharedPreferences 동시 저장) ========
    public void setDong(String dong) {
        this.Dong = dong;
        save("dong", dong);
    }

    public void setHo(String ho) {
        this.Ho = ho;
        save("ho", ho);
    }

    public void setUserName(String userName) {
        this.UserName = userName;
        save("user_name", userName);
    }

    public void setCel(String cel) {
        this.Cel = cel;
        save("cel", cel);
    }

    public void setID(String id) {
        this.ID = id;
        save("id", id);
    }

    public void setIsDriver(boolean isDriver) {
        this.IsDriver = isDriver;
        if (sharedPreferences != null) {
            sharedPreferences.edit().putBoolean("is_driver", isDriver).apply();
        }
    }

    public void setUserId(String userId) {
        this.userId = userId;
        save("user_id", userId);
    }

    public void setOpenMINOR(ArrayList<LobbyOpenData> openMINOR) {
        this.OpenMINOR = openMINOR;
        if (sharedPreferences == null || openMINOR == null) return;
        try {
            JSONArray arr = new JSONArray();
            for (LobbyOpenData data : openMINOR) {
                JSONObject obj = new JSONObject();
                obj.put("minor", data.getMinor());
                obj.put("rssi",  data.getRssi());
                obj.put("id",    data.getId());
                obj.put("dong",  data.getDong());
                obj.put("ho",    data.getHo());
                arr.put(obj);
            }
            sharedPreferences.edit().putString("open_minor", arr.toString()).apply();
        } catch (JSONException ignore) {}
    }

    private void save(String key, String value) {
        if (sharedPreferences != null) {
            sharedPreferences.edit().putString(key, value).apply();
        }
    }
}

File: android/app/src/main/java/com/pms_parkin_mobile/service/UserDataSingleton.kt
package com.pms_parkin_mobile.service

import android.content.Context
import android.content.SharedPreferences
import com.pms_parkin_mobile.dto.LobbyOpenData
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class UserDataSingleton private constructor() {
    private var Dong: String? = null
    private var Ho: String? = null
    private var UserName: String? = null
    private var Cel: String? = null
    private var ID: String? = null
    var isIsDriver: Boolean = false
        private set
    var userId: String? = null
        get() = field ?: sharedPreferences?.getString("user_id", null)
        set(value) {
            field = value
            save("user_id", value)
        }
    private var OpenMINOR: ArrayList<LobbyOpenData>?
    var bigDataSend: Boolean? = null
        get() = field == null || field == true
        set(bigDataSend) {
            field = bigDataSend == null || bigDataSend
        }

    private var sharedPreferences: SharedPreferences? = null

    init {
        OpenMINOR = ArrayList<LobbyOpenData>()
    }

    /** App.init() 에서 호출 — 프로세스 재시작 후에도 데이터 복원  */
    fun init(context: Context) {
        sharedPreferences = context.getApplicationContext()
            .getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        loadFromPrefs()
    }

    private fun loadFromPrefs() {
        if (sharedPreferences == null) return
        Dong = sharedPreferences!!.getString("dong", null)
        Ho = sharedPreferences!!.getString("ho", null)
        UserName = sharedPreferences!!.getString("user_name", null)
        Cel = sharedPreferences!!.getString("cel", null)
        ID = sharedPreferences!!.getString("id", null)
        this.isIsDriver = sharedPreferences!!.getBoolean("is_driver", false)
        userId = sharedPreferences!!.getString("user_id", null)

        val jsonStr = sharedPreferences!!.getString("open_minor", null)
        if (jsonStr != null) {
            try {
                OpenMINOR = ArrayList<LobbyOpenData>()
                val arr = JSONArray(jsonStr)
                for (i in 0..<arr.length()) {
                    val obj = arr.getJSONObject(i)
                    val data = LobbyOpenData()
                    data.minor = obj.optString("minor")
                    data.rssi = obj.optString("rssi")
                    data.id = obj.optString("id")
                    data.dong = obj.optString("dong")
                    data.ho = obj.optString("ho")
                    OpenMINOR!!.add(data)
                }
            } catch (e: JSONException) {
                OpenMINOR = ArrayList<LobbyOpenData>()
            }
        }
    }

    // ======== Getters ========
    fun getDong(): String? {
        return Dong
    }

    fun getHo(): String? {
        return Ho
    }

    fun getUserName(): String? {
        return UserName
    }

    fun getCel(): String? {
        return Cel
    }

    fun getID(): String? {
        return ID
    }

//    fun getUserId(): String? {
//        return userId
//    }

    fun getOpenMINOR(): ArrayList<LobbyOpenData> {
        return OpenMINOR!!
    }

    // ======== Setters (SharedPreferences 동시 저장) ========
    fun setDong(dong: String?) {
        this.Dong = dong
        save("dong", dong)
    }

    fun setHo(ho: String?) {
        this.Ho = ho
        save("ho", ho)
    }

    fun setUserName(userName: String?) {
        this.UserName = userName
        save("user_name", userName)
    }

    fun setCel(cel: String?) {
        this.Cel = cel
        save("cel", cel)
    }

    fun setID(id: String?) {
        this.ID = id
        save("id", id)
    }

    fun setIsDriver(isDriver: Boolean) {
        this.isIsDriver = isDriver
        if (sharedPreferences != null) {
            sharedPreferences!!.edit().putBoolean("is_driver", isDriver).apply()
        }
    }

//    fun setUserId(userId: String?) {
//        this.userId = userId
//        save("user_id", userId)
//    }

    fun setOpenMINOR(openMINOR: ArrayList<LobbyOpenData>?) {
        this.OpenMINOR = openMINOR
        if (sharedPreferences == null || openMINOR == null) return
        try {
            val arr = JSONArray()
            for (data in openMINOR) {
                val obj = JSONObject()
                obj.put("minor", data.minor)
                obj.put("rssi", data.rssi)
                obj.put("id", data.id)
                obj.put("dong", data.dong)
                obj.put("ho", data.ho)
                arr.put(obj)
            }
            sharedPreferences!!.edit().putString("open_minor", arr.toString()).apply()
        } catch (ignore: JSONException) {
        }
    }

    private fun save(key: String?, value: String?) {
        if (sharedPreferences != null) {
            sharedPreferences!!.edit().putString(key, value).apply()
        }
    }

    companion object {
        private const val PREF_NAME = "user_data_pref"

        val instance: UserDataSingleton = UserDataSingleton()
    }
}
